eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php

```diff
@@ -1160,4 +1174,36 @@ public function bulkRemoveTranslation($languageId, $actions)
         ;
         $query->execute();
     }
+
+    /**
+     * Cleanup obsolete lang_mask entries for the given Location.
+     *
+     * @param int $locationId
+     * @param int[] $newLanguageIds
+     * @param bool $alwaysAvailable
+     */
+    public function cleanupTranslationsAfterPublish($locationId, $newLanguageIds, $alwaysAvailable)
+    {
+        $connection = $this->dbHandler->getConnection();
+        /** @var \Doctrine\DBAL\Connection $connection */
+        $query = $connection->createQueryBuilder();
+        $query
+            ->select('id', 'lang_mask', 'parent', 'text_md5')
+            ->from($this->table)
+            ->where('action = :action')
+            ->setParameter('action', 'eznode:' . $locationId);
+
+        $statement = $query->execute();
+
+        // remove specific languages from a bit mask
+        foreach ($statement->fetchAll() as $row) {
+            $oldLanguageIds = $this->languageMaskGenerator->extractLanguageIdsFromMask((int) $row['lang_mask']);
+            foreach ($oldLanguageIds as $oldLanguageId) {
+                if (!in_array($oldLanguageId, $newLanguageIds)) {
```

Comment:
I'm aware this is not very optimal, however I don't see a way to do it using query (could be done via multiple queries which is even less optimal). The problem is that record can contain composite mask (many languages) or there could be multiple single language records to be archived.

Suggestions welcomed :)